### ISSUE TITLE
Outgoing and incoming edges in undirected graphs

### ISSUE THREAD

**Issue by jkinable** [jgrapht/jgrapht#298]
A number of algorithms,  defined on both directed and undirected graphs, iterate over the outgoing or incoming edges of a graph. This creates some problems, because in an `UndirectedGraph`, you cannot iterate over the `outgoingEdgesOf(V vertex)`. Instead, you have to iterate over `edgesOf(V vertex)`. A number of implementations (e.g. `BidirectionalDijkstraShortestPath`) try to resolve this issue as follows:

```
abstract class Specifics{
            public abstract Set<? extends E> edgesOf(V vertex);
        }

        class DirectedSpecifics extends Specifics{

            private DirectedGraph<V, E> graph;

            public DirectedSpecifics(DirectedGraph<V, E> g){
                graph = g;
            }

            @Override
            public Set<? extends E> edgesOf(V vertex){
                return graph.outgoingEdgesOf(vertex);
            }
        }

        class UndirectedSpecifics extends Specifics{
            private Graph<V, E> graph;

            public UndirectedSpecifics(Graph<V, E> g){
                graph = g;
            }

            @Override
            public Set<E> edgesOf(V vertex){
                return graph.edgesOf(vertex);
            }
        }
```

This is becoming a bit of a pattern throughout the library, so we probably want to address this. I don't immediately have a clean solution (requires some more thought so determine the consequences of different solutions). Here are 4 different possibilities:
1. Add default functions outgoingEdgesOf and incomingEdgesOf to the Graph interface, which implements the above behavior.
2. Add a dedicated method to the utility class Graphs.
3. Add this behavior to the UndirectedGraph interface only (this would only solve the problem partially)
4. Create a special GraphDelegator implementation. Perhaps the easiest, least invasive, but probably also the least intuitive one.

Opinions are welcome. Alternatively, propose a solution through a pull request and we'll have a look at it.

**Comment by jsichi** [comment]
As an additional wrinkle, at least one algorithm (see org.jgrapht.alg.isomorphism.GraphOrdering) wants to this in both directions.

**Comment by jsichi** [comment]
I would vote for a variant of 3.

5.  Move the Specifics pattern above to top level (in package org.jgrapht) but rename the interface to something like EdgeAccessor.  Put the factory method in Graphs and make it take a direction parameter to account for the wrinkle noted above (with a corresponding third implementation).

**Comment by d-michail** [comment]
I think I like @jsichi 's suggestion. 

- Option 4 alone is going to be difficult to find by a user of the library. 
- Option 3 alone is not a complete solution. 
- Option 1 will cause more confusion on simpler use-cases than solve things.

However I would also adapt the interface a little bit to return an `Iterable<E>` and not a `Set<E>`. I think this is a bit of a misdesign in the Graph interface (difficult to fix now without breaking things), which is probably since `Iterable<?>` was introduced in Java 5 .

**Comment by jkinable** [comment]
I'll try to work on this and propose something somewhere in the next few weeks... I spent a little too much time on the 1.0.0 release, so need to prioritize some other work first.

**Comment by jsichi** [comment]
It's true that JGraphT started before Java 5, so no Iterable at that time.  But I don't think that's the reason we went with Set (instead of say Collection).  The collections of vertices and edges are indeed sets (the same vertex can't appear twice in the same graph), so the interface should reflect that (similar to how in the Map interface, keySet is a Set, whereas values is a Collection).

**Comment by d-michail** [comment]
@jsichi I agree with the set. However, it would probably be better to return the set as an Iterable. 

For example in method edgesOf() on directed graphs, the current implementation does a copy of the incoming and outgoing edge sets into one set  and then returns it.

If the return type was Iterable, this copy could have been avoided by creating a custom iterator which iterates first over the first set (outgoing edges) and then over the second (incoming edges).

**Comment by jsichi** [comment]
I don't think anything prevents us from creating a custom implementation of the Set interface if we want to eliminate the copying.

BTW for the new EdgeAccessor we're contemplating here, I'm fine with making it produce an Iterable rather than a Set, since the intention is to provide a way to enumerate edges during traversal (as opposed to something like edgesOf, which is intended to define the data structure).

**Comment by jbduncan** [comment]
Have you had a look at [Guava's new graph API](https://google.github.io/guava/releases/20.0-rc1/api/docs/com/google/common/graph/Graph.html)? It looks really well thought out to me, so it may prove to be a source of inspiration for solving this problem.

**Comment by d-michail** [comment]
Solved in #366.
